src/main/java/org/opensearch/knn/index/query/NestedFilteredFloatKNNScorer.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query;
+
+
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.BytesRef;
+import org.opensearch.knn.index.SpaceType;
+import org.opensearch.knn.index.codec.util.KNNVectorSerializer;
+import org.opensearch.knn.index.codec.util.KNNVectorSerializerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+
+/**
+ * Inspired by DiversifyingChildrenFloatKnnVectorQuery in lucene
+ * https://github.com/apache/lucene/blob/7b8aece125aabff2823626d5b939abf4747f63a7/lucene/join/src/java/org/apache/lucene/search/join/DiversifyingChildrenFloatKnnVectorQuery.java#L162
+ *
+ * The class is used in KNNWeight to score filtered KNN field by iterating filtered docs using doc value.
+ * Only the top score and child with top score is returned per parent.
+ */
+public class NestedFilteredFloatKNNScorer implements FilteredKNNScorer {
+	private final float[] queryVector;
+	private final int[] filterIdsArray;
+	private final BinaryDocValues values;
+	private final SpaceType spaceType;
+	private final BitSet parentBitSet;
+	private int currentPos = 0;
+	private int currentParent = -1;
+	private int bestChild = -1;
+	private float currentScore = Float.NEGATIVE_INFINITY;
+
+	public NestedFilteredFloatKNNScorer(
```

Comment:
To pass object, it will be same signature as doExactSearch method.
`doExactSearch(final LeafReaderContext leafReaderContext, final int[] filterIdsArray)`

I think it is better to pass only required parameters instead of passing big object with many other unnecessary data in this case.